zcash_client_backend/src/sync.rs

```diff
@@ -0,0 +1,399 @@
+        match scan_ranges.first() {
+            Some(scan_range) if scan_range.priority() == ScanPriority::Verify => {
+                // Download the blocks in `scan_range` into the block source,
+                // overwriting any existing blocks in this range.
+                let block_meta =
+                    download_blocks(client, fsblockdb_root, db_cache, scan_range).await?;
+
+                // Scan the downloaded blocks and check for scanning errors that
+                // indicate the wallet's chain tip is out of sync with blockchain
+                // history.
+                let scan_ranges_updated =
+                    scan_blocks(params, fsblockdb_root, db_cache, db_data, scan_range)?;
+
+                // Delete the now-scanned blocks, because keeping the entire chain
+                // in CompactBlock files on disk is horrendous for the filesystem.
+                block_deletions.push(delete_cached_blocks(fsblockdb_root, block_meta));
+
+                if scan_ranges_updated {
+                    // The suggested scan ranges have been updated, so we re-request.
+                    scan_ranges = db_data.suggest_scan_ranges()?;
+                } else {
+                    // At this point, the cache and scanned data are locally
+                    // consistent (though not necessarily consistent with the
+                    // latest chain tip - this would be discovered the next time
+                    // this codepath is executed after new blocks are received) so
+                    // we can break out of the loop.
+                    break;
+                }
+            }
+            _ => {
+                // Nothing to verify; break out of the loop
+                break;
+            }
+        }
+    }
+
+    // 7) Loop over the remaining suggested scan ranges, retrieving the requested data
+    //    and calling `scan_cached_blocks` on each range.
+    let scan_ranges = db_data.suggest_scan_ranges()?;
+    debug!("Suggested ranges: {:?}", scan_ranges);
+    for scan_range in scan_ranges.into_iter().flat_map(|r| {
+        // Limit the number of blocks we download and scan at any one time.
+        (0..).scan(r, |acc, _| {
+            if acc.is_empty() {
+                None
+            } else if let Some((cur, next)) = acc.split_at(acc.block_range().start + batch_size) {
+                *acc = next;
+                Some(cur)
+            } else {
+                let cur = acc.clone();
+                let end = acc.block_range().end;
+                *acc = ScanRange::from_parts(end..end, acc.priority());
+                Some(cur)
+            }
+        })
+    }) {
+        // Download the blocks in `scan_range` into the block source.
+        let block_meta = download_blocks(client, fsblockdb_root, db_cache, &scan_range).await?;
+
+        // Scan the downloaded blocks.
+        let scan_ranges_updated =
+            scan_blocks(params, fsblockdb_root, db_cache, db_data, &scan_range)?;
+
```

Comment:
I have this pattern in my sync implementation as well, and it's painfully obvious when looking at the resource manager that sync time takes about twice as long as it needs to because resource utilization keeps switching between bandwidth and CPU. If we could be downloading the next batch _while_ scanning the prior batch, we could better utilize both resources in parallel and get the job done much faster.

Reply:
Indeed, this is exactly what the Android and Swift mobile SDKs do. The approach here was intentionally simplistic for the codebase I wrote it for, but I would want this to also be improved to follow the pipelined approach of the mobile SDKs.

Reply:
I've updated my own sync implementation to be pipelined, limiting resource usage while keeping the CPU busy. I'll be happy to contribute to this PR (or after it) if those changes are welcome.